src/main/scala/is/hail/check/Gen.scala

```diff
@@ -306,11 +306,13 @@ object Gen {
   def zip[T1](g1: Gen[T1]): Gen[T1] = g1
 
   def zip[T1, T2](g1: Gen[T1], g2: Gen[T2]): Gen[(T1, T2)] = Gen { (p: Parameters) =>
-    (g1(p), g2(p))
+    val s = partition(p.rng, p.size, 2)
+    (g1.resize(s(0))(p), g2.resize(s(1))(p))
   }
 
   def zip[T1, T2, T3](g1: Gen[T1], g2: Gen[T2], g3: Gen[T3]): Gen[(T1, T2, T3)] = Gen { (p: Parameters) =>
-    (g1(p), g2(p), g3(p))
+    val s = partition(p.rng, p.size, 3)
+    (g1.resize(s(0))(p), g2.resize(s(1))(p), g3.resize(s(2))(p))
   }
```

Comment:
I think it's less error prone because there is explicit discussion of the `Parameters` `p`. It also seems better encapsulated because we can change the implementation of `Gen` without changing this code.

The first one also avoids explicitly talking about the indices of the partitions array (yay!), unfortunately Scala's type system is too weak to realize that the ordering of the types is preserved.